### PR TITLE
fix(profiling): make sure lineno is always an int and funcname is never None

### LINKED ISSUE #3046: TypeError: '<' not supported between instances of 'NoneType' and 'str'
### Infos
We're using DD to monitor our systems. From time to time we see the following errors popping up in our Sentry environment.

Maybe related to #3019?

### Which version of dd-trace-py are you using?
0.56.0

### Which version of pip are you using?
latest

### What is the result that you get?
``` 
TypeError: '<' not supported between instances of 'NoneType' and 'str'
  File "ddtrace/profiling/scheduler.py", line 53, in flush
    exp.export(events, start, self._last_export)
  File "ddtrace/profiling/exporter/http.py", line 150, in export
    profile = super(PprofHTTPExporter, self).export(events, start_time_ns, end_time_ns)
  File "ddtrace/profiling/exporter/pprof.pyx", line 630, in ddtrace.profiling.exporter.pprof.PprofExporter.export
  File "ddtrace/profiling/exporter/pprof.pyx", line 314, in ddtrace.profiling.exporter.pprof._PprofConverter._build_profile
```

### PR DESCRIPTION
**Pull request by jd**
In some cases, f_lineno can be `None` which would trigger an error in the pprof
output. This is not even caught by mypy, and I've proposed a fix here:

  https://github.com/python/typeshed/pull/6769

This patch makes sure we always use 0 if the line number is `None`.

It also makes sure a funcname is always passed — in practice it's now the case,
but make sure typing reflects that.

Fixes #3046

### CODE REVIEW

File: ddtrace/profiling/collector/_traceback.pyx [hunk 1]
```diff
@@ -12,7 +12,8 @@ cpdef traceback_to_frames(traceback, max_nframes):
         if nframes < max_nframes:
             frame = tb.tb_frame
             code = frame.f_code
-            frames.insert(0, (code.co_filename, frame.f_lineno, code.co_name))
+            lineno = 0 if frame.f_lineno is None else frame.f_lineno
```
Comment (brettlangdon) [inline ddtrace/profiling/collector/_traceback.pyx:15]: `frame.f_lineno or 0`?

File: ddtrace/profiling/collector/threading.py [hunk 2]
```diff
@@ -71,7 +82,10 @@ def __init__(self, wrapped, recorder, tracer, max_nframes, capture_sampler, endp
         self._self_endpoint_collection_enabled = endpoint_collection_enabled
         frame = sys._getframe(2 if WRAPT_C_EXT else 3)
         code = frame.f_code
-        self._self_name = "%s:%d" % (os.path.basename(code.co_filename), frame.f_lineno)
+        self._self_name = "%s:%d" % (
```
Comment (brettlangdon) [inline ddtrace/profiling/collector/threading.py:85]: how is this name used? would `%r` be more useful? (my guess is "no")

File: ddtrace/profiling/exporter/pprof.pyx [hunk 3]
```diff
@@ -151,20 +151,16 @@ class _PprofConverter(object):
...
-        funcname: typing.Optional[str] = None,
+        funcname: str,
     ) -> pprof_LocationType:
         try:
             return self._locations[(filename, lineno, funcname)]
         except KeyError:
-            if funcname is None:
-                real_funcname = "<unknown function>"
-            else:
-                real_funcname = funcname
```
Comment (brettlangdon) [inline ddtrace/profiling/exporter/pprof.pyx:162]: is this related?

**Review by P403n1x87 — COMMENTED** [review]
Because unfortunately the `f_lineno` attribute can be `0` at times, the "correct" way of computing the line number is via `f_lasti` and `f_code.co_lnotab`. The details are in https://github.com/python/cpython/blob/main/Objects/lnotab_notes.txt. The annoying bit is that I believe the formula described there has changed a couple of times across releases.

File: ddtrace/profiling/collector/_traceback.pyx (hunk 1 above)
Comment (jd) [reply ddtrace/profiling/collector/_traceback.pyx:15]: I do prefer to be explicit around None here, not just using any "false" value.

File: ddtrace/profiling/collector/threading.py (hunk 2 above)
Comment (jd) [reply ddtrace/profiling/collector/threading.py:85]: It's shown as the lock name in the UI.

(This is unrelated to the PR though, there's no change here)

File: ddtrace/profiling/exporter/pprof.pyx (hunk 3 above)
Comment (jd) [reply ddtrace/profiling/exporter/pprof.pyx:162]: Yes; we don't want to allow `None` in `FrameType` nor anywhere to be sure we're not going to compare `None` to a `str` at some point when `sort()`ing.

In practice this case does not happen anymore since we always have a funcname since we removed the old memory profiler (based on tracemalloc).

**Comment by jd** [comment]
> Because unfortunately the `f_lineno` attribute can be `0` at times, the "correct" way of computing the line number is via `f_lasti` and `f_code.co_lnotab`. The details are in https://github.com/python/cpython/blob/main/Objects/lnotab_notes.txt. The annoying bit is that I believe the formula described there has changed a couple of times across releases.

That's something we could improve in the future, though it does not impact this PR AFAICT.

**Review by P403n1x87 — APPROVED** [review]

**Review by brettlangdon — APPROVED** [review]
